### PR TITLE
dev/core#2450 - Update source/target contacts on contribution updates

### PR DESCRIPTION
**Pull request by mflandorfer**
Overview
----------------------------------------
https://lab.civicrm.org/dev/core/-/issues/2450

Update source/target activity contacts whenever a contribution is updated.

Before
----------------------------------------
Since PR #19202 CiviCRM would skip updates to source/target activity contact IDs when a contribution is updated, to prevent overwriting them with less accurate contact IDs. However, if a contribution is moved to another contact, the associated activity would still be stuck with the old contact.

After
----------------------------------------
When a contribution is updated and ...

- **a source and target contact exist**, the _target_ contact will be set to the contribution's `contact_id`. The source contact remains untouched.
- **only a source contact exists**, the _source_ contact will be set to the contribution's `contact_id`. This would be the case, if the contribution was originally created via the API.

Technical Details
----------------------------------------

Comments
----------------------------------------

### CODE REVIEW

**Comment by colemanw** [comment]
@civicrm-builder add to whitelist.
@mflandorfer thanks for the PR and extra thanks for including a test!

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -539,6 +540,22 @@ public static function create(&$params) {
...
+          ['record_type_id:name', '=', 'Activity Targets'],
+        ])->execute();
+
+        if ($targetContactQuery->rowCount === 0) {
+          // If no target contact exists, update source contact
+          $activityParams['source_contact_id'] = $contribution->contact_id;
+        }
+        else {
+          // If a target contact exists, update it
+          $activityParams['target_contact_id'] = $contribution->contact_id;
```
Comment (MegaphoneJon) [inline CRM/Contribute/BAO/Contribution.php:556]: On line 541 above, `$activityParams['target_contact_id']` accepts an array; here, it's passed a scalar.  Are both correct, or should this be an array?

File: CRM/Contribute/BAO/Contribution.php [hunk 2]
```diff
@@ -539,6 +540,22 @@ public static function create(&$params) {
         $activityParams['source_contact_id'] = (int) ($params['source_contact_id'] ?? (CRM_Core_Session::getLoggedInContactID() ?: $contribution->contact_id));
         $activityParams['target_contact_id'] = ($activityParams['source_contact_id'] === (int) $contribution->contact_id) ? [] : [$contribution->contact_id];
       }
+      else {
+        // Check if target contact exists
```
Comment (MegaphoneJon) [inline CRM/Contribute/BAO/Contribution.php:544]: The comments on lines 537-539 should be removed if we're now setting target contacts on update.

**Review by MegaphoneJon — CHANGES_REQUESTED** [review]
I just reviewed the code, but haven't done an `r-run`.

I'll first say that this is incredible quality for someone who's never submitted a PR to Civi core before - or frankly, for someone who has.  Very readable, uses modern Civi conventions, and a well-written test.

I spent some time worrying about the comment directly above this patch that says, "Don't update the activity target values when the contribution is updated" but I can't see a scenario where this patch doesn't create the correct values.

Part of me wonders if this update should happen in the `movecontrib` extension - but tbh I think that functionality belongs in core anyway.

File: CRM/Contribute/BAO/Contribution.php (hunk 1 above)
Comment (mflandorfer) [reply CRM/Contribute/BAO/Contribution.php]: Thanks for your feedback!

Yes, I think this should be an array.

**Comment by MegaphoneJon** [comment]
Many thanks @mflandorfer, my colleague is setting up to do a manual test now.  The manual test plus my code review should be enough to get this approved for merge.

**Comment by megaphonetech-dennis** [comment]
I just finished a manual test. Everything worked great. And, I didn't find anything odd or broken as a result of applying this patch.

File: tests/phpunit/CRM/Contribute/BAO/ContributionTest.php [hunk 3]
```diff
@@ -1688,4 +1688,91 @@ public function testCancelOrderWithPledge(): void {
...
+    ];
+
+    // Case 1: Only source contact, no target contact
+
+    $contribution = $this->callAPISuccess('Contribution', 'create', array_merge(
+      $contributionParams,
+      ['contact_id' => $contactId_1]
+    ))['values'][0];
+
+    $activity = $this->callAPISuccess('Activity', 'get', [
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contribute/BAO/ContributionTest.php:1713]: You don't need to change this but one small shortcut is 

```
$activity = $this->callAPISuccessGetSingle('Activity', ['source_record_id' => $contribution['id']])
```

of course you have to set a limit if there could be legimately more than one...

File: CRM/Contribute/BAO/Contribution.php [hunk 4]
```diff
@@ -533,12 +534,25 @@ public static function create(&$params) {
...
       if (!$activityParams['id']) {
-        // Don't set target contacts on update as these will have been
-        // correctly created and we risk overwriting them with
-        // 'best guess' params.
         $activityParams['source_contact_id'] = (int) ($params['source_contact_id'] ?? (CRM_Core_Session::getLoggedInContactID() ?: $contribution->contact_id));
         $activityParams['target_contact_id'] = ($activityParams['source_contact_id'] === (int) $contribution->contact_id) ? [] : [$contribution->contact_id];
       }
+      else {
+        // Check if target contact exists
+        $targetContactQuery = ActivityContact::get(FALSE)->setLimit(1)->setWhere([
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php:542]: So this is obviously a high-volume function & we need to be mindful of queries. In particular unnecessary updates can be locking queries - so I would do a bit more work here to check whether we are actually making a change of the contribution->contact is already the target - probably by opening up ```['record_type_id:name', '=', 'Activity Targets'],``` to

```['record_type_id:name', 'IN', ['Activity Targets', 'Activity Sources']],```

and doing extra checks in the php layer. If it starts to get a bit long then extract it

**Comment by eileenmcnaughton** [comment]
I agree with @MegaphoneJon - this looks really good. I think it's worth doing a bit more work to avoid doing updates if there is no change though as this is a high-volume function

**Comment by eileenmcnaughton** [comment]
Test Result (3 failures / +3)CRM_Contribute_BAO_ContributionTest.testUpdateActivityContactOnContributionContactChangeCRM_Contribute_Form_Contribution_ConfirmTest.testPayNowPaymentapi_v3_SyntaxConformanceTest.testCreateSingleValueAlter with data set #82
--

**Comment by eileenmcnaughton** [comment]
This looks good - are you able to squash the commits before we merge?

**Comment by pfigel** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@mflandorfer I just checked in on this again & it is now in conflict :-(

Can you ping me once it's fixed so I hopefully don't miss it again

**Comment by mflandorfer** [comment]
@eileenmcnaughton I just resolved the merge conflict

**Comment by eileenmcnaughton** [comment]
@mflandorfer jenkins is still not loving it - can you do a git pull --rebase with a force push in the hope it cheers up?

**Comment by mflandorfer** [comment]
@eileenmcnaughton OK, looks good now

**Comment by eileenmcnaughton** [comment]
@mflandorfer yay - let's lock it in!

**Pull request merged**